src/directLineStreaming.ts

```diff
@@ -0,0 +1,351 @@
+// In order to keep file size down, only import the parts of rxjs that we use
+
+import { BehaviorSubject } from 'rxjs/BehaviorSubject';
+import { Observable } from 'rxjs/Observable';
+import { Subscriber } from 'rxjs/Subscriber';
+import * as BFSE from 'botframework-streaming';
+
+import {
+  Activity,
+  ConnectionStatus,
+  Conversation,
+  IBotConnection,
+  Media,
+  Message
+} from './directLine';
+
+const DIRECT_LINE_VERSION = 'DirectLine/3.0';
+const MAX_RETRY_COUNT = 3;
+const refreshTokenLifetime = 30 * 60 * 1000;
+const timeout = 20 * 1000;
+const refreshTokenInterval = refreshTokenLifetime / 2;
+
+interface DirectLineStreamingOptions {
+  token?: string,
+  conversationId?: string,
+  domain?: string,
+  // Attached to all requests to identify requesting agent.
+  botAgent?: string
+}
+
+class StreamHandler implements BFSE.RequestHandler {
+  private connectionStatus$;
+  private subscriber: Subscriber<Activity>;
+  private shouldQueue: () => boolean;
+  private activityQueue: Array<Activity> = [];
+
+  constructor(s: Subscriber<Activity>, c$: Observable<ConnectionStatus>, sq: () => boolean) {
+    this.subscriber = s;
+    this.connectionStatus$ = c$;
+    this.shouldQueue = sq;
+  }
+
+  public setSubscriber(s: Subscriber<Activity>) {
+    this.subscriber = s;
+  }
+
+  async processRequest(request: BFSE.IReceiveRequest, logger?: any): Promise<BFSE.StreamingResponse> {
+    const streams = [...request.streams];
+    const stream0 = streams.shift();
+    const activitySetJson = await stream0.readAsString();
+    const activitySet = JSON.parse(activitySetJson);
+
+    if (activitySet.activities.length !== 1) {
+      // Only one activity is expected in a set in streaming
+      this.subscriber.error(new Error('there should be exactly one activity'));
+      const r = new BFSE.StreamingResponse();
+      r.statusCode = 500;
```

Comment:
StreamingResponse should perhaps take a status code or signal to derive a status code internally.

`new BFSE.SteamingResponse({ type: 'error' })`
`new BFSE.StreamingResponse({ code: 500 })`

Reply:
BFSE.StreamingResponse has a static method called create. I've changed the code to use that.